### PR TITLE
UCP/bcopy: Code refactoring. Handle all datatypes in one place

### PR DESCRIPTION
**Pull request by shssf**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2747/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/156/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2748/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2132/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2749/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/157/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2134/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2135/ for details (Mellanox internal link).

File: src/ucp/tag/eager_snd.c [hunk 1]
```diff
@@ -48,30 +60,30 @@ static size_t ucp_tag_pack_eager_dt(void *dest, void *arg)
     ucp_request_t *req = arg;
     size_t length;
 
-    length         = req->send.length;
```
Comment (yosefe) [inline src/ucp/tag/eager_snd.c]: IMHO we need to rename ucp_tag_pack_eager_dt() to ucp_tag_pack_eager_**single**_dt()

File: src/ucp/tag/eager_snd.c [hunk 2]
```diff
@@ -48,30 +60,30 @@ static size_t ucp_tag_pack_eager_dt(void *dest, void *arg)
...
-    length               = req->send.length;
     hdr->super.super.tag = req->send.tag;
     hdr->req.sender_uuid = req->send.ep->worker->uuid;
     hdr->req.reqptr      = (uintptr_t)req;
 
     ucs_assert(req->send.state.offset == 0);
-    ucp_tag_pack_eager_dt_copy(hdr + 1, req->send.buffer, &req->send.state,
-                               length, req->send.datatype);
-    return sizeof(*hdr) + length;
+    return sizeof(*hdr) + ucp_tag_pack_eager_dt_copy(hdr + 1, req->send.buffer,
```
Comment (yosefe) [inline src/ucp/tag/eager_snd.c]: we should rename this to ucp_tag_pack_eager_sync_**single**_dt() and assert the returned length is == req->send.length

File: src/ucp/tag/eager_snd.c [hunk 3]
```diff
@@ -258,6 +180,9 @@ static ucs_status_t ucp_tag_eager_bcopy_single(uct_pending_req_t *self)
                                                  ucp_tag_pack_eager_dt);
     if (status == UCS_OK) {
         ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+        if (UCP_DATATYPE_GENERIC == (req->send.datatype & UCP_DATATYPE_CLASS_MASK)) {
```
Comment (yosefe) [inline src/ucp/tag/eager_snd.c]: shall we add inline function ucp_request_dt_finish() to contain these 3 lines (which also repeat in ucp_tag_eager_bcopy_multi and ucp_tag_eager_sync_bcopy_single)

**Comment by yosefe** [comment]
@shssf pls check how this change affects ucp eager bcopy performance

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2756/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2142/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/172/ for details.

**Comment by yosefe** [comment]
bot:bgateNEW:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/187/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/190/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/191/ for details.

**Comment by yosefe** [comment]
@shamisp pls take a look

**Comment by shamisp** [comment]
👍

**Pull request merged**